scripts/validate_changelog_utils.py

```diff
@@ -0,0 +1,108 @@
+        endIndex = version_name.find(str_after_version)
+        if endIndex == -1:
+            continue
+        version_name = version_name[:endIndex]
+        if pattern.fullmatch(version_name) != None:
+            stable_versions[version_name] = versions[version]
+    return stable_versions
+
+def should_skip_changelog(response_json):
+    if "labels" in response_json:
+        pr_labels = response_json["labels"]
+        for label in pr_labels:
+            if label["name"] == "skip changelog":
+                return True
+    return False
+
+def check_has_changelog_diff(diff):
+    changelog_diff_matches = re.search(changelog_diff_regex, diff, re.MULTILINE)
+    if not changelog_diff_matches:
+        raise Exception("Add a non-empty changelog entry in a CHANGELOG.md or add a `skip changelog` label if not applicable.")
+
+def parse_contents_url(files_response_json):
+    for file_json in files_response_json:
+        if file_json["filename"] == changelog_filename:
+            return file_json["contents_url"]
+    raise Exception("No CHANGELOG.md file in PR files")
+
+def extract_added_lines(diff):
+    changelog_diff_matches = re.search(changelog_diff_regex, diff, re.MULTILINE)
+    diff_starting_at_changelog = diff[changelog_diff_matches.end():]
+    first_changelog_diff = re.search(diff_file_start_regex, diff_starting_at_changelog, re.MULTILINE)
+    first_changelog_diff_index = 0
+    if first_changelog_diff:
+        first_changelog_diff_index = first_changelog_diff.end()
+    last_reachable_index = diff_starting_at_changelog.find(any_diff_substring, first_changelog_diff_index)
+    if last_reachable_index == -1:
+        last_reachable_index = len(diff_starting_at_changelog)
+    diff_searchable = diff_starting_at_changelog[first_changelog_diff_index:last_reachable_index]
+
+    diff_lines = diff_searchable.split('\n')
+    return list(map(remove_plus, list(filter(is_line_added, diff_lines))))
+
+def check_contains_pr_link(added_lines):
+    for added_line in added_lines:
+        pr_link_matches = re.search(pr_link_regex, added_line)
+        if not pr_link_matches:
+            raise Exception("The changelog entry \"" + added_line + "\" should contain a link to the original PR that matches `" + pr_link_regex + "`")
+
+def check_version_section(content, added_lines):
+    lines = content.split("\n")
+    versions = group_by_versions(lines)
+    unreleased_group = extract_unreleased_group(versions)
+    stable_versions = extract_stable_versions(versions)
+
+    for added_line in added_lines:
+        if added_line not in unreleased_group:
+            raise Exception(added_line + " should be placed in 'Unreleased' section")
+
+        for stable_version in stable_versions:
+            if added_line in stable_versions[stable_version]:
+                raise Exception("The changelog entry \"" + added_line + "\" is already contained in " + stable_version + " changelog.")
```

Comment:
Does it mean that I won't be able to change changelog of released version?
If yes, maybe it's worth adding a "modify-old-changelog" label to allow modification of old changelog? It can be useful if you want to fix a typo or clarify something after a question from a customer? What do you think?

Reply:
You're right. Makes sense, however, I don't want to overcomplicate it. If you specifically make a PR that clarifies an old entry, you just disable the check. This check is mostly introduced for two reasons: 1) so that we don't forget to update CHANGELOG.md; 2) so that our lines do not appear in random places after rebase after release.
I want to keep the system as simple as possible. For special cases it's absolutely OK to check turn the check off.